File: java/dp/house_robber.java
class Solution {
    public int rob(int[] nums) {
        // problem: maximize the sum (no conseq)
        // subproblem: rob or not + suffix
        // subpr count #n
        // choices: 2
        // solution O(n)
        int len = nums.length -1;
        int [] memo = new int[len+1];
        Arrays.fill(memo, -1);
        return dpUtil(0, len, memo, nums);

    }

    public int dpUtil(int from, int to, int[] memo, int [] nums){
        if(from > to){
            return 0;
        }
        if(memo[from] == -1){
            if(from == to){
                memo[from] = nums[from];
            }
            else{
                memo[from] = Math.max(dpUtil(from+1,to,memo,nums),nums[from]+dpUtil(from+2, to, memo, nums));
            }
        }
        
        return memo[from];

    }
}

File: python/1d_dynamic_programming/house_robber.py
from typing import List
class Solution:
    def rob(self, nums: List[int]) -> int:
        size = len(nums)
        cache = dict()

        def dfs(i):
            if i >= size:
                return 0
            if i in cache:
                return cache[i]

            rob = nums[i] + dfs(i+2)
            skip = dfs(i+1)

            cache[i] = max(rob,skip)
            return cache[i]

        return dfs(0)